### PR TITLE
Added test cases for 4 CSharp11 features

### PR DESCRIPTION
**Pull request by cristian-ambrosini-sonarsource**

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/DisposeNotImplementingDisposeTest.cs [hunk 1]
```diff
@@ -35,11 +35,21 @@ public class DisposeNotImplementingDisposeTest
 
         [TestMethod]
         public void DisposeNotImplementingDispose_CSharp9() =>
-            builder.AddPaths("DisposeNotImplementingDispose.CSharp9.Part1.cs", "DisposeNotImplementingDispose.CSharp9.Part2.cs").WithTopLevelStatements().Verify();
+            builder.AddPaths("DisposeNotImplementingDispose.CSharp9.Part1.cs", "DisposeNotImplementingDispose.CSharp9.Part2.cs")
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/DisposeNotImplementingDisposeTest.cs:38]: There's no value in splitting existing UTs that fit on one line into multiple lines.

**Comment by csaba-sagi-sonarsource** [comment]
You should add test cases for DeadStores that document the behaviour of empty MultilneRawStringLiterals and InterpolatedRawStringLiterals.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/DoNotDecreaseMemberVisibility.CSharp11.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
+﻿using System;
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/DoNotDecreaseMemberVisibility.CSharp11.cs:1]: Lets have a call tomorrow to go into the details, both DoNotDecreaseMemberVisibility and DoNotHideBaseClassMethods test cases are not valid.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/DoNotLockOnSharedResource.CSharp11.cs [hunk 3]
```diff
@@ -5,13 +5,11 @@ namespace Tests.Diagnostics
 {
     public class LockOnThisOrType
     {
-        public void MyLockingMethod()
+        public void RawStringLiterals()
         {
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/DoNotLockOnSharedResource.CSharp11.cs:9]: DoNotLockWeakIdentityObjects is missing similar test cases with raw string literals.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/DoNotShiftByZeroOrIntSize.CSharp11.cs [hunk 4]
```diff
@@ -0,0 +1,10 @@
+﻿using System;
+
+void UnsignedRightShiftOperator(int i, ulong q)
+{
+    i = i >>> 60; // FN
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/DoNotShiftByZeroOrIntSize.CSharp11.cs:5]: Also add test cases with  `>>>=`

**Comment by csaba-sagi-sonarsource** [comment]
For DoNotTestThisWithIsOperator, you can add something like: 
```
    public class SomeClass: ArrayList
    {
        public void SomeClass()
        {
            var a = this is [1, 2, 3];
        }
    }
```

**Comment by csaba-sagi-sonarsource** [comment]
For DotNotOverloadOperatorEqual, we should not raise if the operator == is overloaded if it comes from implementing interface IEqualityOperators

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EventHandlerDelegateShouldHaveProperArguments.CSharp11.cs [hunk 5]
```diff
@@ -0,0 +1,31 @@
+﻿using System;
+
+namespace Tests.Diagnostics
+{
+    public interface IMyInterface
+    {
+        public static event EventHandler SomeStaticEvent;
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EventHandlerDelegateShouldHaveProperArguments.CSharp11.cs:7]: Please add test cases with virtual and abstract events.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ExceptionShouldNotBeThrownFromUnexpectedMethods.CSharp11.cs [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+    interface IMyInterface : IDisposable
+    {
+        static event EventHandler OnSomeArrow
+        {
+            add => throw new Exception(); // Noncompliant
+            remove => throw new InvalidOperationException(); // Compliant - allowed exception
+        }
+    }
+
+    class MyClass : IMyInterface
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ExceptionShouldNotBeThrownFromUnexpectedMethods.CSharp11.cs:14]: Why this test case, and in general why implementing IDisposable?

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/DoNotShiftByZeroOrIntSizeTest.cs [hunk 7]
```diff
@@ -25,18 +25,35 @@ namespace SonarAnalyzer.UnitTest.Rules
...
+            builder.AddPaths("DoNotShiftByZeroOrIntSize.cs").Verify();
 
 #if NET
+
         [TestMethod]
         public void DoNotShiftByZeroOrIntSize_CSharp9() =>
-            OldVerifier.VerifyAnalyzerFromCSharp9Console(@"TestCases\DoNotShiftByZeroOrIntSize.CSharp9.cs", new DoNotShiftByZeroOrIntSize());
+            builder.AddPaths("DoNotShiftByZeroOrIntSize.CSharp9.cs")
+                .WithTopLevelStatements()
+                .WithOptions(ParseOptionsHelper.FromCSharp9)
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/DoNotShiftByZeroOrIntSizeTest.cs:40]: WithTopLevelStatements already enforces the usage of Csharp9, so no need to add an option specifically for that.

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/DoNotShiftByZeroOrIntSizeTest.cs [hunk 8]
```diff
@@ -25,18 +25,35 @@ namespace SonarAnalyzer.UnitTest.Rules
...
         public void DoNotShiftByZeroOrIntSize_CSharp10() =>
-            OldVerifier.VerifyAnalyzerFromCSharp10Library(@"TestCases\DoNotShiftByZeroOrIntSize.CSharp10.cs", new DoNotShiftByZeroOrIntSize());
+            builder.AddPaths("DoNotShiftByZeroOrIntSize.CSharp10.cs")
+                .WithOptions(ParseOptionsHelper.FromCSharp10)
+                .Verify();
+
+        [TestMethod]
+        public void DoNotShiftByZeroOrIntSize_CSharp11() =>
+            builder.AddPaths("DoNotShiftByZeroOrIntSize.CSharp11.cs")
+                .WithTopLevelStatements()
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/DoNotShiftByZeroOrIntSizeTest.cs:52]: You can move the test cases into a class and remove the usage of top level statements.

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/EmptyStatementTest.cs [hunk 9]
```diff
@@ -18,39 +18,55 @@
...
 #if NET
+
         [TestMethod]
         public void EmptyStatement_CodeFix_CSharp9() =>
-            OldVerifier.VerifyCodeFix<EmptyStatementCodeFix>(
-                @"TestCases\EmptyStatement.CSharp9.cs",
-                @"TestCases\EmptyStatement.CSharp9.Fixed.cs",
-                new EmptyStatement(),
-                ParseOptionsHelper.FromCSharp9,
-                OutputKind.ConsoleApplication);
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/EmptyStatementTest.cs:53]: Have you checked if the OutputKind was important for the test case before? I see you have committed it.

**Review by csaba-sagi-sonarsource — CHANGES_REQUESTED** [review]
Looks good, andI have left a few comments.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/DoNotDecreaseMemberVisibility.CSharp11.cs (hunk 2 above)
Comment (csaba-sagi-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/DoNotDecreaseMemberVisibility.CSharp11.cs:1]: Sorry for the confusion. They make sense!

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ExceptionShouldNotBeThrownFromUnexpectedMethods.CSharp11.cs (hunk 6 above)
Comment (cristian-ambrosini-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ExceptionShouldNotBeThrownFromUnexpectedMethods.CSharp11.cs:14]: I think it was left from the CSharp9 copied file,  my bad.

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/DoNotLockWeakIdentityObjectsTest.cs [hunk 10]
```diff
@@ -26,12 +26,24 @@ namespace SonarAnalyzer.UnitTest.Rules
     [TestClass]
     public class DoNotLockWeakIdentityObjectsTest
     {
+        private readonly VerifierBuilder builderCS = new VerifierBuilder<CS.DoNotLockWeakIdentityObjects>();
+        private readonly VerifierBuilder builderVB = new VerifierBuilder<VB.DoNotLockWeakIdentityObjects>();
         [TestMethod]
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/DoNotLockWeakIdentityObjectsTest.cs:31]: Missing empty line above

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/DoNotLockWeakIdentityObjectsTest.cs [hunk 11]
```diff
@@ -26,12 +26,24 @@ namespace SonarAnalyzer.UnitTest.Rules
     [TestClass]
     public class DoNotLockWeakIdentityObjectsTest
     {
+        private readonly VerifierBuilder builderCS = new VerifierBuilder<CS.DoNotLockWeakIdentityObjects>();
+        private readonly VerifierBuilder builderVB = new VerifierBuilder<VB.DoNotLockWeakIdentityObjects>();
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/DoNotLockWeakIdentityObjectsTest.cs:30]: No need for the VB field as the builder is only used in 1 test.

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/EmptyStatementTest.cs [hunk 12]
```diff
@@ -18,39 +18,53 @@
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
 
+using Microsoft.CodeAnalysis.CSharp;
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/EmptyStatementTest.cs:21]: Unused using

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EventHandlerDelegateShouldHaveProperArguments.CSharp11.cs (hunk 5 above)
Comment (csaba-sagi-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EventHandlerDelegateShouldHaveProperArguments.CSharp11.cs]: Actually with abstract events I think it is not possible to make a test case however with virtual it is. Here you go:
```
public interface IMyInterface
    {
        public static virtual event EventHandler VirtualEvent;
        public static void VirtualMethod<T>(object sender, EventArgs e) where T : IMyInterface
        {
            T.VirtualEvent.Invoke(null, null);            // Noncompliant
            VirtualEvent.Invoke(null, null);            // Noncompliant
            T.VirtualEvent.Invoke(null, e);            // Compliant
            VirtualEvent.Invoke(null, e);            // Compliant
        }
    }
```

**Review by csaba-sagi-sonarsource — APPROVED** [review]
LGTM!
